retis/src/core/filters/meta/filter.rs

```diff
@@ -136,118 +129,426 @@ impl MetaLoad {
+                                lhs: Box::new(lhs),
+                                op: infix.clone(),
+                                rhs: Box::new(Self::parse_primary(inner)?),
+                            }
+                        }
+                        _ => bail!("while parsing the expression"),
+                    }
+                }
+            }
+            _ => bail!("unexpected rule while parsing expression"),
+        }
+
+        Ok(lhs)
+    }
+}
+
+#[derive(Clone, Debug)]
+enum BooleanOp {
+    And,
+    Or,
+}
+
+#[derive(Clone, Debug)]
+enum AstNode {
+    RelOpExpr {
+        lhs: Lhs,
+        op: RelOp,
+        rhs: Rhs,
+    },
+    BooleanExpr {
+        lhs: Box<AstNode>,
+        op: BooleanOp,
+        rhs: Box<AstNode>,
+    },
+}
+
+struct MetaExpr<'a> {
+    filter: eBpfProg,
+    btf_info: &'a BtfInfo,
+    btf: &'a Btf,
+    btf_type: Type,
+    offt: u32,
+}
+
+impl<'a> MetaExpr<'a> {
+    fn init_filter(store_arg: bool) -> eBpfProg {
+        let mut filter = eBpfProg::new();
+
+        if store_arg {
+            filter.add(eBpfInsn::mov(MovInfo::Reg {
+                src: BpfReg::R1,
+                dst: BpfReg::R0,
+            }));
+
+            filter.exit_retval_eq(0);
+
+            // Argument address (any walkable type).
+            // The assumption is that this remains untouched.
+            filter.add(eBpfInsn::mov(MovInfo::Reg {
+                src: BpfReg::R1,
+                dst: BpfReg::R6,
+            }));
+        }
+
+        // R7 will be used as new base in each
+        // iteration.
+        filter.add(eBpfInsn::mov(MovInfo::Reg {
+            src: BpfReg::R6,
+            dst: BpfReg::R7,
+        }));
+
+        filter
     }
 
-    fn emit_load_ptr(offt: u32, mask: u64) -> Result<MetaOp> {
-        let mut op: MetaOp = MetaOp::new();
-        op.l.offt = u16::try_from(offt / 8)?;
-        op.l.r#type = PTR_BIT;
-        op.l.mask = mask;
+    fn new(btf_info: &'a BtfInfo, sym: &str, arg: bool) -> Result<MetaExpr<'a>> {
+        let types = btf_info
+            .resolve_types_by_name(sym)
+            .map_err(|e| anyhow!("unable to resolve {sym} data type {e}"))?;
+
+        let (btf, r#type) = match types.iter().find(|(_, t)| matches!(t, Type::Struct(_))) {
+            Some(r#struct) => r#struct,
+            None => bail!("Could not resolve {sym} to a struct"),
+        };
 
-        Ok(op)
+        Ok(Self {
+            filter: Self::init_filter(arg),
+            btf_info,
+            btf,
+            btf_type: r#type.clone(),
+            offt: 0,
+        })
     }
 
-    fn emit_load(btf: &Btf, r#type: &Type, offt: u32, bfs: u32, mask: u64) -> Result<MetaOp> {
-        let mut op: MetaOp = MetaOp::new();
-        let lop = op.load_ref_mut();
-        let mut t = r#type.clone();
-        let mut type_iter = btf.type_iter(
-            r#type
+    fn check_expr(&self, field: &LhsNode, rel_op: RelOp, rval: Rhs) -> Result<XlateCtx> {
```

Comment:
The function name could be more descriptive. My understanding is it is constructing a context, not only validating the expression.

Reply:
I would not name it against the fact that it returns a context, though (looks a little redundant). 
`finalize_expr()` ?
Feel free to suggest a better name